edb/server/dbview/dbview.pyx

```diff
@@ -218,18 +233,81 @@ cdef class Database:
             conn = await self.tenant.acquire_pgcon(self.name)
             try:
                 await g.execute(conn, self)
+            except Exception as e:
+                logger.warning("Failed to persist function cache",
+                               exc_info=True)
+                continue
             finally:
                 self.tenant.release_pgcon(self.name, conn)
 
-            for _, units in ops:
+            for query_req, units in ops:
                 units.cache_state = CacheState.Present
+                if self._active_tx_list:
+                    # Any active tx would delay the time we flip to func cache
+                    units.tx_seq_id = self._tx_seq
+                    self._func_cache_gt_tx_seq[query_req] = units
+                else:
+                    units[0].maybe_use_func_cache()
                 self._cache_notify_queue.put_nowait(str(units[0].cache_key))
 
+    cdef inline int tx_seq_begin_tx(self):
+        self._tx_seq += 1
+
+        if self._tx_seq == TX_SEQ_MAX:
+            # Wrap around to a valid ID, and append a marker to the list
+            self._tx_seq = 1
+            self._active_tx_list[TX_SEQ_MAX] = None
+            # Also append a wrapping marker to the pending func cache list
+            self._func_cache_gt_tx_seq[None] = None
+
+        self._active_tx_list[self._tx_seq] = None
+        return self._tx_seq
+
+    cdef inline tx_seq_end_tx(self, int seq):
+        if seq not in self._active_tx_list:
+            return
+
+        # First, remove the ending transaction from the active list
+        self._active_tx_list.pop(seq, None)
+
+        # Second, grab the seq ID of the oldest active transaction
+        if self._active_tx_list:
+            seq = next(iter(self._active_tx_list.keys()))
+            if seq == TX_SEQ_MAX:
+                # Hit a wrapping marker, pop it out so that we can see the real
+                # oldest active transaction later
+                self._active_tx_list.pop(TX_SEQ_MAX)
+
+        # Stop early if we don't have func cache to activate
+        if not self._func_cache_gt_tx_seq:
+            return
+
+        # If all tx ended, we should just activate all pending func cache
+        if not self._active_tx_list:
+            for units in self._func_cache_gt_tx_seq.values():
+                units[0].maybe_use_func_cache()
+            self._func_cache_gt_tx_seq.clear()
+            return
+
+        # At last, keep activating func cache until the oldest active tx
+        drops = []
+        for query_req, units in self._func_cache_gt_tx_seq.items():
+            if query_req is None:
+                # Hit a wrapping marker, grab the real oldest active tx ID
+                seq = next(iter(self._active_tx_list.keys()))
+                drops.append(None)
+            elif units.tx_seq_id < seq:
+                units[0].maybe_use_func_cache()
+                drops.append(query_req)
+            else:
+                break
```

Comment:
Instead of waiting for every finish transaction to drop before we start using the function, could we track the transaction id in the cache and start using it in later transactions even while old ones are still pending?

I think it probably doesn't matter *that* much for real-world use cases, but it will play a lot nicer with an idea for testing I want to do.

The testing idea is that we can have a test mode that runs each query test twice, with the idea that it will hopefully hit the cache on the second loop, and I think all the extra transactions are likely to cause some trouble.

Reply:
Actually, maybe that test idea won't work that well anyway, since it often won't have been written in time for the second loop anyway...

Maybe the strategy should be to run a full test suite, then run the full suite a second time now that the cache is populated.

Reply:
> Instead of waiting for every finish transaction to drop before we start using the function, could we track the transaction id in the cache and start using it in later transactions even while old ones are still pending?

Yeah, I thought about this. Without changing the QueryUnit data structure, I think I can also patch pgcon.pyx to take an additional param to use func call.

> Maybe the strategy should be to run a full test suite, then run the full suite a second time now that the cache is populated.

This sounds more reasonable to me.